File: Year 2/Semester 4/MPP/ProtobuffProjectCSharp/ProtobuffNetworking/ProtoWorker.cs
using System;
using System.Threading;
using System.Net.Sockets;

using ProtobuffProject.services;
using GRPCProject.Protocol;
using Google.Protobuf;
using ProtobuffProject.domain;

namespace ProtobuffNetworking
{
	public class ProtoWorker : IObserver
	{
		private IServices server;
		private TcpClient connection;

		private NetworkStream stream;
		private volatile bool connected;

		public ProtoWorker(IServices server, TcpClient connection)
		{
			this.server = server;
			this.connection = connection;
			try
			{

				stream = connection.GetStream();
				connected = true;
			}
			catch (Exception e)
			{
				Console.WriteLine(e.StackTrace);
			}
		}

		public virtual void Run()
		{
			while (connected)
			{
				try
				{
					Request request = Request.Parser.ParseDelimitedFrom(stream);
					Response response = handleRequest(request);
					if (response != null)
					{
						sendResponse(response);
					}
				}
				catch (Exception e)
				{
					Console.WriteLine(e.StackTrace);
				}

				try
				{
					Thread.Sleep(1000);
				}
				catch (Exception e)
				{
					Console.WriteLine(e.StackTrace);
				}
			}

			try
			{
				stream.Close();
				connection.Close();
			}
			catch (Exception e)
			{
				Console.WriteLine("Error " + e);
			}
		}

		private Response handleRequest(Request request)
		{
			Response response = null;
			Request.Types.Type reqType = request.Type;
			switch (reqType)
			{
				case Request.Types.Type.Login:
				{
					Console.WriteLine("Login request ...");
					Employee<int> employee = ProtoUtils.GetEmployee(request);
					try
					{
						lock (server)
						{
							return ProtoUtils.CreateLoginResponse(
								server.AuthenticateEmployee(employee, this));
						}
					}
					catch (MyException e)
					{
						connected = false;
						return ProtoUtils.CreateErrorResponse(e.Message);
					}
				}
				case Request.Types.Type.Logout:
				{
					Console.WriteLine("Logout request");
					Employee<int> employee = ProtoUtils.GetEmployee(request);
					try
					{
						lock (server)
						{
							server.Logout(employee);
						}

						connected = false;
						return ProtoUtils.CreateOkResponse();

					}
					catch (MyException e)
					{
						return ProtoUtils.CreateErrorResponse(e.Message);
					}
				}
				case Request.Types.Type.BuyTicket:
				{
					Console.WriteLine("Buy ticket request");
					Ticket<int> ticket = ProtoUtils.GetTicket(request);
					try
					{
						lock (server)
						{
							server.AddTickets(ticket);
						}

						return ProtoUtils.CreateOkResponse();
					}
					catch (MyException e)
					{
						return ProtoUtils.CreateErrorResponse(e.Message);
					}
				}
				case Request.Types.Type.GetAllShows:
				{
					Console.WriteLine("Get all shows request");
					try
					{
						lock (server)
						{
							return ProtoUtils.CreateGetAllShowsResponse(server.GetAllShowsDTO());
						}
					}
					catch (MyException e)
					{
						return ProtoUtils.CreateErrorResponse(e.Message);
					}
				}
				case Request.Types.Type.GetAllArtists:
				{
					Console.WriteLine("Get all artists request");
					try
					{
						lock (server)
						{
							return ProtoUtils.CreateGetAllArtistsResponse(server.GetAllArtists());
						}
					}
					catch (MyException e)
					{
						return ProtoUtils.CreateErrorResponse(e.Message);
					}
				}
				case Request.Types.Type.FindArtist:
				{
					Console.WriteLine("Find artist request");
					try
					{
						lock (server)
						{
							return ProtoUtils.CreateFindArtistResponse(
								server.FindArtist(ProtoUtils.GetIdArtist(request)));
						}
					}
					catch (MyException e)
					{
						return ProtoUtils.CreateErrorResponse(e.Message);
					}
				}
			}

			return response;
		}

		private void sendResponse(Response response)
		{
			Console.WriteLine("sending response " + response);
			lock (stream)
			{
				response.WriteDelimitedTo(stream);
				stream.Flush();
			}
		}

		public void TicketAdded(Ticket<int> ticket)
		{
			Console.WriteLine("Ticket added " + ticket);
			try
			{
				sendResponse(ProtoUtils.CreateTicketAddedResponse(ticket));
			}
			catch (Exception e)
			{
				Console.WriteLine(e.StackTrace);
			}
		}
	}

}

File: Year 2/Semester 4/Programming and Projection Environments/Week 10/ProtobuffProjectJavaORM/ProtobuffNetworking/src/main/java/gRPCProject/networking/protobuffprotocol/ProtoWorker.java
package gRPCProject.networking.protobuffprotocol;

import gRPCProject.domain.Employee;
import gRPCProject.domain.Ticket;
import gRPCProject.service.IObserver;
import gRPCProject.service.IService;
import gRPCProject.service.MyException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;


public class ProtoWorker implements Runnable, IObserver {
    private final IService server;
    private final Socket connection;
    private InputStream input;
    private OutputStream output;
    private volatile boolean connected;
    private static final Logger logger = LogManager.getLogger();

    public ProtoWorker(IService server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try {
            output = connection.getOutputStream();
            input = connection.getInputStream();
            connected = true;
            logger.info("Worker created");
        } catch (IOException e) {
            logger.error(e);
        }
    }

    public void run() {
        while (connected) {
            try {
                System.out.println("Waiting requests ...");
                logger.info("Waiting requests ...");
                Protobuffs.Request request = Protobuffs.Request.parseDelimitedFrom(input);
                System.out.println("Request received: " + request);
                logger.info("Request received: " + request);
                Protobuffs.Response response = handleRequest(request);
                if (response != null) {
                    sendResponse(response);
                }
            } catch (IOException e) {
                logger.error(e);
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                logger.error(e);
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
            logger.info("Worker closed");
        } catch (IOException e) {
            logger.error(e);
            System.out.println("Error " + e);
        }
    }

    public void ticketAdded(Ticket ticket) {
        System.out.println("Ticket sold " + ticket);
        logger.info("Ticket sold " + ticket);
        try {
            sendResponse(ProtoUtils.createTicketAddedResponse(ticket));
        } catch (IOException e) {
            logger.error(e);
        }
    }

    private Protobuffs.Response handleRequest(Protobuffs.Request request) {
        Protobuffs.Response response = null;
        switch (request.getType()) {
            case Login -> {
                System.out.println("Login request ...");
                logger.info("Login request ...");
                Employee employee = ProtoUtils.getEmployee(request);
                try {
                    logger.info("Employee authenticated");
                    return ProtoUtils.createLoginResponse(server.authenticateEmployee(employee, this));
                } catch (MyException e) {
                    connected = false;
                    logger.error(e);
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case Logout -> {
                System.out.println("Logout request");
                logger.info("Logout request");
                Employee employee = ProtoUtils.getEmployee(request);
                try {
                    server.logout(employee, this);
                    connected = false;
                    logger.info("Employee logged out");
                    return ProtoUtils.createOkResponse();

                } catch (MyException e) {
                    logger.error(e);
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case BUY_TICKET -> {
                System.out.println("Buy ticket request");
                logger.info("Buy ticket request");
                Ticket ticket = ProtoUtils.getTicket(request);
                try {
                    server.addTickets(ticket);
                    logger.info("Ticket added");
                    return ProtoUtils.createOkResponse();
                } catch (MyException e) {
                    logger.error(e);
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case GET_ALL_SHOWS -> {
                System.out.println("Get all shows request");
                logger.info("Get all shows request");
                try {
                    return ProtoUtils.createGetAllShowsResponse(server.getAllShowsDTO());
                } catch (MyException e) {
                    logger.error(e);
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case GET_ALL_ARTISTS -> {
                System.out.println("Get all artists request");
                logger.info("Get all artists request");
                try {
                    return ProtoUtils.createGetAllArtistsResponse(server.getAllArtists());
                } catch (MyException e) {
                    logger.error(e);
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case FIND_ARTIST -> {
                System.out.println("Find artist request");
                logger.info("Find artist request");
                int id = ProtoUtils.getIdArtist(request);
                try {
                    return ProtoUtils.createFindArtistResponse(server.findArtist(id));
                } catch (MyException e) {
                    logger.error(e);
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
        }
        return response;
    }

    private void sendResponse(Protobuffs.Response response) throws IOException {
        System.out.println("sending response " + response);
        logger.info("sending response " + response);
        response.writeDelimitedTo(output);
        output.flush();
    }
}
